File: Python/0103-binary-tree-zigzag-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# Time:  O(n)
# Space: O(n)
class Solution:
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []
        result = []
        prev = [root]
        while prev:
            cur = []
            vals = []
            for node in prev:
                vals.append(node.val)
                if node.left:
                    cur.append(node.left)
                if node.right:
                    cur.append(node.right)
            prev = cur
            # or vals.reverse()
            result.append(vals[::-1] if len(result) % 2 else vals)
        return result

File: C++/0103-binary-tree-zigzag-level-order-traversal.cpp
// Tags: Amazon Tree BFS
// Time:  O(n)
// Space: O(n)
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<TreeNode*> cur {root};
        vector<vector<int>> res;
        if (!root)
            return res;
        while (!cur.empty()) {
            vector<int> vals;
            vector<TreeNode*> next;
            for (auto &node: cur) {
                vals.push_back(node->val);
                if (node->left)
                    next.push_back(node->left);
                if (node->right)
                    next.push_back(node->right);
            }
            cur = next;
            if (res.size() % 2 != 0)
                reverse(vals.begin(), vals.end());
            res.push_back(vals);
        }
        return res;
    }
};
